sklearn/metrics/cluster/_supervised.py

```diff
@@ -138,6 +145,209 @@ def contingency_matrix(labels_true, labels_pred, *, eps=None, sparse=False):
+        Ground truth class labels to be used as a reference
+
+    labels_pred : array, shape = [n_samples]
+        Cluster labels to evaluate
+
+    Returns
+    -------
+    nested list [[D00, D01], [D10, D11]].
+
+    Examples
+    --------
+
+    Perfectly matching labelings have all non-zero entries on the
+    diagonal regardless of actual label values:
+
+      >>> from sklearn.metrics.cluster import pair_confusion_matrix
+      >>> pair_confusion_matrix([0, 0, 1, 1], [0, 0, 1, 1])
+      [[8, 0], [0, 4]]
+      >>> pair_confusion_matrix([0, 0, 1, 1], [1, 1, 0, 0])
+      [[8, 0], [0, 4]]
+
+    Labelings that assign all classes members to the same clusters
+    are complete but may be not always pure, hence penalized, and
+    have some off-diagonal non-zero entries:
+
+      >>> pair_confusion_matrix([0, 0, 1, 2], [0, 0, 1, 1])  # doctest: +ELLIPSIS
+      [[8, 2], [0, 2]]
+
+    The matrix is not symmetric:
+
+      >>> pair_confusion_matrix([0, 0, 1, 1], [0, 0, 1, 2])  # doctest: +ELLIPSIS
+      [[8, 0], [2, 2]]
+
+    If classes members are completely split across different clusters, the
+    assignment is totally incomplete, hence the matrix has all zero
+    diagonal entries:
+
+      >>> pair_confusion_matrix([0, 0, 0, 0], [0, 1, 2, 3])
+      [[0, 0], [12, 0]]
+
+    References
+    ----------
+
+    .. [Hubert1985] L. Hubert and P. Arabie, Comparing Partitions,
+      Journal of Classification 1985
+      https://link.springer.com/article/10.1007%2FBF01908075
+
+    See also
+    --------
+    adjusted_pair_confusion_matrix: Adjusted Rand Score
+    adjusted_mutual_info_score: Adjusted Mutual Information
+
+    """
+    labels_true, labels_pred = check_clusterings(labels_true, labels_pred)
+    n_samples = np.int64(labels_true.shape[0])
+    n_classes = np.unique(labels_true).shape[0]
+    n_clusters = np.unique(labels_pred).shape[0]
+
+    # Special limit cases: no clustering since the data is not split;
+    # or trivial clustering where each document is assigned a unique cluster.
+    # These are perfect matches hence all pairs are counted either under
+    # D00 or D11.
+    if n_classes == n_clusters == 1:
+        return [[0,0], [0,n_samples * (n_samples - 1)]]
+    elif n_classes == n_clusters == 0:
+        return [[0,0], [0,0]]
+    elif n_classes == n_clusters == n_samples:
+        return [[n_samples * (n_samples - 1),0], [0,0]]
+
+    # Computation using the contingency data
+    contingency = contingency_matrix(labels_true, labels_pred, sparse=True, dtype=np.int64)
+    n_c = np.ravel(contingency.sum(axis=1))
+    n_k = np.ravel(contingency.sum(axis=0))
+    #rows,cols = contingency.nonzero()
+    #D11 = sum(n_ij*(n_ij-1) for n_ij in contingency.data)
+    Sum_Squares = sum(n_ij*n_ij for n_ij in contingency.data)
+    D11 = Sum_Squares - n_samples
+    #D01 = sum(contingency[c,k] * (n_k[k] - contingency[c,k]) for c,k in zip(rows,cols))
+    D01 = sum(contingency.dot(n_k)) - Sum_Squares
```

Comment:
Is this a marked improvement than the implementation of the pairwise confusion in adjusted_rand_index using `_comb2`?

Reply:
Yes, for larger clusterings it runs quite a bit faster. A _comb2 implementation of the un-adjusted rand score might be:
```
    sum_comb = sum(_comb2(n_ij) for n_ij in contingency.data)
    sum_comb_c = sum(_comb2(n_c) for n_c in np.ravel(contingency.sum(axis=1)))
    sum_comb_k = sum(_comb2(n_k) for n_k in np.ravel(contingency.sum(axis=0)))
    D11 = 2 * sum_comb
    D01 = 2 * (sum_comb_k-sum_comb)
    D10 = 2 * (sum_comb_c-sum_comb)
    D00 = n_samples * n_samples - n_samples - D01 - D10 - D11
```
A runtime test case can be generated with:
```
n = 10000
clustering1 = list(range(n))
clustering2 = list(range(n))
# randomly create consecutive pairs for 5% of the entries
np.random.seed(12345)
clustering1_pairs = sorted(np.random.choice(clustering1[1:], size=int(0.05*n), replace=False))
for i in clustering1_pairs:
    clustering1[i-1] = clustering1[i]
clustering2_pairs = sorted(np.random.choice(clustering2[1:], size=int(0.05*n), replace=False))
for i in clustering2_pairs:
    clustering2[i-1] = clustering2[i]
```
On my machine %timeit for the two implementation returns:
```
pair_confusion_matrix(clustering1, clustering2)
10.3 ms ± 530 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
pair_confusion_matrix_comb2(clustering1, clustering2)
20.2 ms ± 735 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```
Run-time differences highly depend on the structure of the clusterings. I have not seen a case where my implementation is slower.